internal/alpini/netty4/alpini-netty4-base/src/main/java/com/linkedin/alpini/netty4/handlers/ConnectionControlHandler.java

```
@@ -32,28 +44,42 @@
  */
 @ChannelHandler.Sharable
 public class ConnectionControlHandler extends ConnectionLimitHandler {
+  private static final Logger LOGGER = LogManager.getLogger(ConnectionControlHandler.class);
   private final LongAdder _activeCount = new LongAdder();
   private final LongAdder _inactiveCount = new LongAdder();
   private final Semaphore _activeSemaphore = new Semaphore(1);
   private final Semaphore _inactiveSemaphore = new Semaphore(1);
 
-  public ConnectionControlHandler(@Nonnegative int connectionLimit) {
-    this(() -> connectionLimit);
-  }
-
-  public ConnectionControlHandler(@Nonnull IntSupplier connectionLimit) {
-    super(connectionLimit);
+  public ConnectionControlHandler(
+      @Nonnull IntSupplier connectionLimit,
+      @Nonnull Consumer<Integer> connectionCountRecorder) {
+    super(connectionLimit, connectionCountRecorder);
   }
 
   @Override
   public void channelActive(ChannelHandlerContext ctx) throws Exception {
     _activeCount.increment();
+    _connectionCountRecorder.accept(getConnectedCount());
 
     Channel parent = ctx.channel().parent();
     if (parent != null && _activeSemaphore.tryAcquire()) {
       parent.eventLoop().submit(() -> {
         if (getConnectedCount() > getConnectionLimit()) {
+          LOGGER.error(
+              "Connection limit exceeded! Active connections: {}, Limit: {}",
+              getConnectedCount(),
+              getConnectionLimit());
           parent.config().setAutoRead(false);
+          Channel clientChannel = ctx.channel(); // Get the current client's channel
+          if (clientChannel.isActive()) {
+            FullHttpResponse response = new DefaultFullHttpResponse(
+                HttpVersion.HTTP_1_1,
+                HttpResponseStatus.TOO_MANY_REQUESTS,
+                Unpooled.copiedBuffer("Connection limit exceeded on Venice routers", StandardCharsets.US_ASCII));
```

can we include the connection limit in the response?

---

nitpick: I feel that we overload 429 too much.

---

We probably don't need to expose internal details for users; 429 should be good enough.
We have a error log about the limit on router side.

---

Yes, since we are aligned with the approach now, I am going to remove all the changes in ConnectionControlHandler.

To more context: ConnectionControlHandler would stall the client request, while ConnectionLimitHander would fail fast. We are going to remove all the changes in ConnectionControlHandler and keep the stalling behavior. We might need it to prevent DDOS attack in future.
For now, by default, Venice will use ConnectionLimitHandler.